### ISSUE TITLE
FR: Ignoring empty YAML properties when duplicated

### ISSUE THREAD

**Issue by robsonsobral** [platers/obsidian-linter#821]
## Is Your Feature Request Related to a Problem? Please Describe.

I guess it isn't a fuuuuuuuuuuuull feature.

Sometimes we import a template data which has YAML properties with the same name as some of the ones already being used. Then, the LINTER throws an error, even if they are empty.

## Describe the Solution You'd Like

I would like that, in case there's a duplicated property, the empty one to be removed before an error being thrown.

**Please include an example where applicable:**

``` yaml
subtitle: This is the subtitle
another_property:
subtitle:
```

``` yaml
subtitle: This is the subtitle
anotherProperty:
```

## Describe Alternatives You've Considered

The only alternative I can think of is to search for the duplicates manually, which is annoying, giving that probably the duplicated properties are out of order.

## Additional Context

Thank you for the useful plugin!

**Comment by pjkaufman** [comment]
This would need to be its own rule. I say this because it would have to remove or comment out the duplicate keys before any YAML parsing is attempted. The YAML parser used does not allow duplicate keys at the same level, so that is why you see the error.

**Comment by robsonsobral** [comment]
> The YAML parser used does not allow duplicate keys at the same level, so that is why you see the error.

Yeah, I know! However, I see no point at reporting an error for something empty.

It would net be a so easy task. The array of duplicated properties would need to trigger a purging process of the empty ones before to throw the error.

**Comment by pjkaufman** [comment]
I am not sure how you are importing duplicate properties, but handling this scenario seems like an edge case to me. If there is a desire to add this from the community, I can work on it. In the meantime, I am more than happy to merge any PRs that address this.

**Comment by robsonsobral** [comment]
It's ease to get into this. Let's say I create a new entry, `foobar.md`, and save it. The frontmatter will be created:

```yaml
---
title: Foobar
tags: []
created: 2023-08-01 21:13:46
modified: 2023-08-01 21:17:14
---
```

Then, I insert data from a template:

![image](https://github.com/platers/obsidian-linter/assets/931642/16f27d71-270c-4cd2-b4bd-fbf00ac80aea)

The template may have its own `title`, `created`, etc properties.

---

I took a look at the source code, to try a PR.

The clean-up should run at the beginning of the function [`loadYAML()`](https://github.com/platers/obsidian-linter/blob/0a9b86621d541c33f2d35a4770c85885b4ae55fd/src/utils/yaml.ts#L77), before to run the `load()` from `js-yaml', right?

**Comment by pjkaufman** [comment]
It depends on what you are trying to achieve with the change you are making. The Linter should allow users to decide whether it will throw an error based on whether or not there are duplicate instances of a key. To achieve this, you will have to do 1 of 2 things (there could be another option I am not thinking of):

1. Add a new rule that removes empty instances of duplicated keya
2. Remove duplicate keys from the YAML in `LoadYAML` before it tries to parse the YAML.

I believe option 1 makes the most sense since it is a lot more maintainable in the long run.

Here is a link to the docs for adding a new rule: https://platers.github.io/obsidian-linter/contributing/adding-a-rule/. The new docs have not yet been fully integrated, but I am working on that.

If you go with it being a rule, it needs to be a special rule that runs before other rules do so it needs to have the `hasSpecialExecutionOrder` value set to true in the constructor. If it helps, it would look something like [format-tags-in-yaml](https://github.com/platers/obsidian-linter/blob/0a9b86621d541c33f2d35a4770c85885b4ae55fd/src/rules/format-tags-in-yaml.ts).

Then the only other thing is to add it to the set of rules that run before the other rules [here](https://github.com/platers/obsidian-linter/blob/0a9b86621d541c33f2d35a4770c85885b4ae55fd/src/rules-runner.ts#L94).

**Comment by robsonsobral** [comment]
> I believe option 1 makes the most sense since it is a lot more maintainable in the long run.

I disagree. It looks more like a necessary step to me, like replacing `\t` with double whitespace. Something that has to be done for the other rules to work. **However**, since you're the person who has to maintain the code in the long run, I'll try to follow option 1!

**Comment by pjkaufman** [comment]
I am under the assumption that you are planning to remove duplicate keys as hiding them from the parser, while feasible, could cause people to assume they have valid YAML when they do not. This is more minor. The bigger problem occurs with YAML rules running against the YAML with duplicate keys. The Linter uses regex to get some values for keys. This will not work when there are duplicate keys. It will grab the first instance of the key and ignore the other. This will create unexpected output and potentially cause errors in the Linter rules.

The tabs issue, on the other hand, is more that YAML does not support tabs and so I just went with swapping them to spaces because it would not affect the output of actual rules. Unfortunately, duplicate keys does not work this way.

**Comment by robsonsobral** [comment]
The [release of the version 1.4.3](https://obsidian.md/changelog/2023-07-31-desktop-v1.4.3/) made this question obsolete:

> Templates: When using the "Insert template" command, properties inside template files will be merged with the current file.

**Comment by pjkaufman** [comment]
Gotcha. I am glad to hear that.
